Replace Alchy with SQLAlchemy

seallard opened the pull request:
## Description
This PR replaces `Alchy` with `SQLAlchemy`. `Alchy` is not maintained and locks us to an old version of `SQLAlchemy`, which is why we want to replace it. Closes https://github.com/Clinical-Genomics/trailblazer/issues/298.

The session/transaction management approach applied here ties the session and transaction scope to that of a request or CLI command. One session and one transaction is used per request/command. 

Once the request/command has been processed, the transaction is committed and the session is removed (unless an exception occurred - then the transaction is rolled back first). I think it makes sense since it is very simple and the requests/commands run quickly (as far as I understand).

This differs from what is done in CG where we limit the scope of the transactions more tightly by doing commits throughout the codebase. I think that approach makes sense in cg since some of the commands run for 10 + hours. We don't want to keep a single transaction open for that long, it would be problematic since transactions can lock tables.

TODOs:
- [x] Add database module
- [x] Replace `Flask-Alchy`
- [x] Replace `AlchyManager`
- [x] Replace `Alchy` in models with `SQLAlchemy`
- [x] Add session/transaction management for CLI and API
- [x] Use global `scoped_session` in handlers
- [x] Remove commits in codebase
- [x] Patch tests
- [x] Add serialization (`Alchy` had built in `to_dict` methods on models)

I'd appreciate some help with the QA testing of this PR.

### Fixed
- Refactor project to only use SQLAlchemy
- Standardize session/transaction management
- Remove Flask-Alchy
- Remove Flask-SQLAlchemy
- Upgrade Click to latest version
- Upgrade Flask to latest version

```
bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_trailblazer -t trailblazer -b  remove-alchy -a
```

### This [version](https://semver.org/) is a
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

seallard commented on requirements.txt:
```diff
@@ -1,4 +1,3 @@
-Alchy
 Click<7
 SQLAlchemy<1.4
```
Will upgrade this along with Python in separate PR:s

seallard commented on trailblazer/store/models.py:
```diff
@@ -23,6 +21,14 @@ class Info(Model):
     created_at = Column(types.DateTime, default=datetime.datetime.now)
     updated_at = Column(types.DateTime)
 
+    def to_dict(self) -> dict:
```
`Alchy` added these `to_dict` methods by default, so they need to be implemented manually now. We do it in a hacky way in cg, but I prefer this explicit way. There are many other options for serialization in Python, for instance the `Marsmallow` package.

I feel like we should standardize which tools we use for these things and document it. Since we are quite few developers compared to the amount of code we maintain, it is good if we settle on using one solution for solving common problems like
- What do we use for serialization?
- What do we use for CLI:s?
- What ORM do we use?
etc.

seallard replied:
Also, let me know if some of the fields should be excluded. Have not looked through them all yet

henrikstranneheim commented on trailblazer/server/api.py:
```diff
@@ -73,13 +74,12 @@ def analyses():
 def analysis(analysis_id):
     """Display a single analysis."""
-    analysis_obj = store.get_analysis_with_id(analysis_id=analysis_id)
+    analysis_obj = store.get_analysis_with_id(analysis_id)
```
```suggestion
    analysis: Optional[Analysis] = store.get_analysis_with_id(analysis_id)
```

henrikstranneheim commented on trailblazer/store/core.py:
Should we also rename this module to store?

henrikstranneheim commented on trailblazer/store/database.py:
```diff
@@ -0,0 +1,39 @@
+
+def initialize_database(db_uri: str) -> None:
+    global SESSION, ENGINE
+    ENGINE = create_engine(db_uri, pool_pre_ping=True)
```
Nice

henrikstranneheim commented on trailblazer/store/database.py:
```diff
@@ -0,0 +1,39 @@
+
+
+def create_all_tables() -> None:
+    """Create all tables in status db."""
```
```suggestion
    """Create all tables in the database."""
```

henrikstranneheim commented on trailblazer/store/database.py:
```diff
@@ -0,0 +1,39 @@
+
+
+def drop_all_tables() -> None:
+    """Drop all tables in status db."""
```
```suggestion
    """Drop all tables in the database."""
```

henrikstranneheim commented on trailblazer/store/database.py:
```diff
@@ -0,0 +1,39 @@
+
+
+def get_tables() -> List[str]:
+    """Get a list of all tables in status db."""
```
```suggestion
    """Get a list of all tables in the database."""
```

henrikstranneheim reviewed (approved):
Beautiful 🤩 !

I can do the testing on stage.

seallard commented on trailblazer/server/api.py:
```diff
@@ -14,6 +14,7 @@
     TrailblazerStatus,
 )
 from trailblazer.server.ext import store
```
This import is broken after removing the `ext` module

seallard commented on trailblazer/cli/core.py:
```diff
@@ -15,12 +16,28 @@
+        session.rollback()
+        print(f"Session commit failed: {commit_exception}", file=sys.stderr)
+    finally:
+        session.remove()
```
Fixed, upgraded `Click` to version 8 which has a nice feature called [`with_resource`](https://click.palletsprojects.com/en/8.1.x/api/#click.Context.with_resource)

henrikstranneheim commented:
### API ✅ 

```

```
<img width="1727" alt="Screenshot 2023-11-13 at 16 24 51" src="https://github.com/Clinical-Genomics/trailblazer/assets/584738/664f5bd5-07a0-4a9b-8d03-ccb06703f825">

<img width="1727" alt="Screenshot 2023-11-13 at 16 25 48" src="https://github.com/Clinical-Genomics/trailblazer/assets/584738/46897765-acdb-433f-a011-b8a5a695c105">

<img width="1727" alt="Screenshot 2023-11-13 at 16 25 24" src="https://github.com/Clinical-Genomics/trailblazer/assets/584738/24af5754-2a48-405e-a925-0ac87e58ad30">

seallard commented:
Deployed to Hasta stage:
```θ66° [sebastian.allard@hasta:~] [S_base] 5s $ trailblazer scan
Error updating analysis for: case - grandaphid : FileNotFoundError
Error updating analysis for: case - goodbengal : MissingSqueueOutput
Error updating analysis for: case - helpedlab : MissingSqueueOutput
Error updating analysis for: case - newgopher : FileNotFoundError
Error updating analysis for: case - onestag : MissingSqueueOutput
Error updating analysis for: case - funnyaphid : FileNotFoundError
Error updating analysis for: case - proudfish : MissingSqueueOutput
Error updating analysis for: case - setcolt : MissingSqueueOutput
Error updating analysis for: case - handyshrimp : FileNotFoundError
Error updating analysis for: case - divinetomcat : MissingSqueueOutput
Error updating analysis for: case - HW2V2DRX2 : FileNotFoundError
Error updating analysis for: case - savingstag : FileNotFoundError
Error updating analysis for: case - topewe : MissingSqueueOutput
Error updating analysis for: case - fullgoat : MissingSqueueOutput
Error updating analysis for: case - bolddrum : FileNotFoundError
Error updating analysis for: case - abovehusky : FileNotFoundError
Error updating analysis for: case - chiefgannet : FileNotFoundError
Error updating analysis for: case - grandoyster : FileNotFoundError
Status in Tower: coolkodiak - 282044 - WPjrjs23yxQwq
Error logging case - coolkodiak :  JSONDecodeError
Updated status amazedbaboon - 282547: completed 
Updated status savedkid - 282549: completed 
Updated status coolbug - 282550: completed 
Updated status merrydoe - 282551: completed 
Updated status newmaggot - 282552: completed 
Updated status steadyfoal - 282553: completed 
Updated status secureosprey - 282554: completed 
Updated status setturtle - 282555: completed 
Updated status crackfowl - 282556: completed 
Updated status masterrat - 282557: completed 
Updated status politerabbit - 282558: completed 
Updated status loyalpony - 282559: completed 
Updated status crispbaboon - 282560: completed 
Updated status rightdory - 282561: completed 
Updated status honestbug - 282562: completed 
Updated status setibex - 282563: completed 
Updated status validgoblin - 282564: completed 
Updated status superbmacaw - 282565: completed 
Updated status justkoi - 282566: completed 
Updated status bigkoala - 282567: completed 
Updated status proadder - 282568: completed 
Updated status crispturkey - 282569: completed 
Updated status mightychow - 282570: completed 
Updated status primedeer - 282571: completed 
Updated status choicecattle - 282572: completed 
Updated status modernamoeba - 282573: completed 
Updated status mastermutt - 282575: completed 
Updated status finekitten - 282576: completed 
Updated status likedcod - 282577: completed 
Updated status handyperch - 282578: completed 
Updated status choiceraven - 282579: completed 
Updated status largesalmon - 282580: completed 
Updated status grandpup - 282581: completed 
Updated status uniquewombat - 282582: completed 
Updated status selectbass - 282583: completed 
Updated status brightwhale - 282584: completed 
Updated status fairmolly - 282585: completed 
Updated status soughtbengal - 282586: completed 
Updated status starbadger - 282587: completed 
Updated status niceelk - 282589: completed 
Updated status eagerthrush - 282591: completed 
Updated status clearmarten - 282592: completed 
Updated status abovebee - 282593: completed 
Updated status restedhen - 282594: completed 
Updated status directteal - 282596: completed 
Updated status chiefweasel - 282597: completed 
Updated status oneshark - 282598: completed 
Updated status fondhawk - 282599: completed 
Updated status finerheron - 282600: completed 
Updated status cutebird - 282601: failed 
Updated status alertfilly - 282602: failed 
Updated status frankamoeba - 282603: completed 
Updated status amazedalien - 282604: completed
```

seallard commented:
The api works as intended
![image](https://github.com/Clinical-Genomics/trailblazer/assets/33596072/ac149176-a3df-466b-b146-d474243a33b3)

seallard commented:
Deployed to stage/prod on vms and hasta.
